lightning/src/ln/channelmanager.rs

```diff
@@ -815,19 +815,73 @@ macro_rules! maybe_break_monitor_err {
+		assert!(res.1.is_ok());
+		res
+	} };
+	($self: ident, $channel_lock: expr, $channel_state: expr, $channel_entry: expr,
+	 $raa: expr, $commitment_update: expr, $order: expr, $chanmon_update: expr,
 	 $pending_forwards: expr, $funding_broadcastable: expr, $funding_locked: expr) => { {
 		let mut htlc_forwards = None;
 		let counterparty_node_id = $channel_entry.get().get_counterparty_node_id();
 
-		{
-			if !$pending_forwards.is_empty() {
+		let res = loop {
+			let forwards: Vec<(PendingHTLCInfo, u64)> = $pending_forwards; // Force type-checking to resolve
+			if !forwards.is_empty() {
 				htlc_forwards = Some(($channel_entry.get().get_short_channel_id().expect("We can't have pending forwards before funding confirmation"),
-					$channel_entry.get().get_funding_txo().unwrap(), $pending_forwards));
+					$channel_entry.get().get_funding_txo().unwrap(), forwards));
 			}
 
+			let chanmon_update: Option<ChannelMonitorUpdate> = $chanmon_update; // Force type-checking to resolve
+			if chanmon_update.is_some() {
+				// On reconnect (or monitor restoration), we, by definition, only resend a
+				// funding_locked if there have been no commitment updates, so the only channel
+				// monitor update which could also be associated with a funding_locked would be
+				// the funding_created/funding_signed monitor update. That monitor update failing
+				// implies that we won't send funding_locked until it's been updated, so we can't
+				// have a funding_locked and a monitor update here (so we don't bother to handle it
+				// correctly below).
+				assert!($funding_locked.is_none());
+				// A channel monitor update makes no sense without either a funding_locked or a
+				// commitment update to process after it. Since we can't have a funding_locked, we
+				// only bother to handle the monitor-update + commitment_update case below.
+				assert!($commitment_update.is_some());
+			}
+
+			if let Some(msg) = $funding_locked {
+				// Similar to the above, this implies that we're letting the funding_locked fly
+				// before it should be allowed to.
+				assert!(chanmon_update.is_none());
+				$channel_state.pending_msg_events.push(events::MessageSendEvent::SendFundingLocked {
+					node_id: counterparty_node_id,
+					msg,
+				});
+				if let Some(announcement_sigs) = $self.get_announcement_sigs($channel_entry.get()) {
+					$channel_state.pending_msg_events.push(events::MessageSendEvent::SendAnnouncementSignatures {
+						node_id: counterparty_node_id,
+						msg: announcement_sigs,
+					});
+				}
+				$channel_state.short_to_id.insert($channel_entry.get().get_short_channel_id().unwrap(), $channel_entry.get().channel_id());
+			}
+
+			let funding_broadcastable: Option<Transaction> = $funding_broadcastable; // Force type-checking to resolve
 			macro_rules! handle_cs { () => {
+				if let Some(monitor_update) = chanmon_update {
+					assert!($order == RAACommitmentOrder::RevokeAndACKFirst);
+					assert!(funding_broadcastable.is_none());
```

Comment:
"According to the spec, if the funding transaction has not been broadcast, the funding node should forget about the channel (and thus not participate in a proposed reestablishment)"?

Reply:
No, we only forget if we haven't received a funding_signed, not just not yet broadcast. I've added a comment, though.